src/Microsoft.AspNet.OData.Shared/Formatter/ClrTypeCache.cs

```diff
@@ -9,19 +9,34 @@ namespace Microsoft.AspNet.OData.Formatter
 {
     internal class ClrTypeCache
     {
-        private ConcurrentDictionary<Type, IEdmTypeReference> _cache =
+        private ConcurrentDictionary<Type, IEdmTypeReference> _clrToEdmTypeCache =
             new ConcurrentDictionary<Type, IEdmTypeReference>();
 
+        private ConcurrentDictionary<IEdmTypeReference, Type> _edmToClrTypeCache = new ConcurrentDictionary<IEdmTypeReference, Type>();
+
         public IEdmTypeReference GetEdmType(Type clrType, IEdmModel model)
         {
             IEdmTypeReference edmType;
-            if (!_cache.TryGetValue(clrType, out edmType))
+            if (!_clrToEdmTypeCache.TryGetValue(clrType, out edmType))
             {
                 edmType = model.GetEdmTypeReference(clrType);
-                _cache[clrType] = edmType;
+                _clrToEdmTypeCache[clrType] = edmType;
             }
 
             return edmType;
         }
+
+        public Type GetClrType(IEdmTypeReference edmType, IEdmModel edmModel)
```

Comment:
>IEdmTypeReference [](start = 31, length = 17)

Should this map to a TypeReference or a Type? How do we determine nullability (or do we care)?